Common/System/OSD.cpp

```diff
@@ -99,6 +113,88 @@ void OnScreenDisplay::ShowAchievementUnlocked(int achievementID) {
 	entries_.insert(entries_.begin(), msg);
 }
 
+void OnScreenDisplay::ShowAchievementProgress(int achievementID, float duration_s) {
+	double now = time_now_d();
+
+	for (auto &entry : sideEntries_) {
+		if (entry.numericID == achievementID && entry.type == OSDType::ACHIEVEMENT_PROGRESS) {
+			// Duplicate, let's just bump the timer.
+			entry.startTime = now;
+			entry.endTime = now + (double)duration_s;
+			// We're done.
+			return;
+		}
+	}
+
+	// OK, let's make a new side-entry.
+	Entry entry;
+	entry.numericID = achievementID;
+	entry.type = OSDType::ACHIEVEMENT_PROGRESS;
+	entry.startTime = now;
+	entry.endTime = now + (double)duration_s;
+	sideEntries_.insert(sideEntries_.begin(), entry);
+}
+
+void OnScreenDisplay::ShowChallengeIndicator(int achievementID, bool show) {
+	double now = time_now_d();
+
+	for (auto &entry : sideEntries_) {
+		if (entry.numericID == achievementID && entry.type == OSDType::ACHIEVEMENT_CHALLENGE_INDICATOR && !show) {
+			// Hide and eventually delete it.
+			entry.endTime = now + (double)FadeoutTime();
+			// Found it, we're done.
+			return;
+		}
+	}
+
+	if (!show) {
+		// Sanity check
+		return;
+	}
+
+	// OK, let's make a new side-entry.
+	Entry entry;
+	entry.numericID = achievementID;
+	entry.type = OSDType::ACHIEVEMENT_CHALLENGE_INDICATOR;
+	entry.startTime = now;
+	entry.endTime = now + 10000000.0;  // Don't auto-fadeout.
```

Comment:
nit: Should probably make this some constexpr and not have the comment.  Though, I think infinity would work fine, not that anyone's gonna leave PPSSPP open for 100+ days...

-[Unknown]

Reply:
I'm a little paranoid about infinities in general (like, accidentally dividing by them in some alpha blend calculation or whatever), but you're right that they'd probably work here. It doesn't really matter though. I'll make a constant.